### ISSUE TITLE
Code release ?

### ISSUE THREAD

**Issue by AjayTalati** [jfutoma/MGP-RNN#1]
Hi, 

it's written in your paper that you'll be releasing the code - may I ask when this will happen?

Thanks

**Comment by jfutoma** [comment]
Sorry for the delay!  I've been away traveling the past month and haven't had time to get around to it.  I have a long flight early next week when I finally should have a lot of time to work on this.  I need to finish the code to simulate data before I release so that you'll be able to run it on something, since unfortunately I won't be able to release our data. Thanks for your patience and interest! :)

**Comment by AjayTalati** [comment]
Hi, thanks very much for the reply 

I've spent quite a while working on your paper and really like it! I also promised some folks I'd work on understanding your code when it's released, and explain both the paper and the code to them. It's really appreciated that you're also including a simulator to generate some test data :+1: 

Hope you have a comfortable flight  :)

All the best, Ajay

**Comment by AjayTalati** [comment]
Hi, I was just wondering whether it's possible to push the code for the paper yet? 

If the simulator for the test data is not ready yet, I could try to write that myself - I've got a simple state space simulator which uses, http://www.statsmodels.org/stable/index.html, so could submit a PR once I get it working?

Thanks a lot for your help, Ajay

**Comment by jfutoma** [comment]
Just dumped some code, will be updating with the simulator soon.  Sorry for the delay!

**Comment by jfutoma** [comment]
Can you give me an estimate of what your sample size / time series size looks like?  Not all of the computation speedup tricks I used are necessarily useful if your data is smaller. Thanks!

**Comment by AjayTalati** [comment]
Hi Joe, wow that's really nice of you - thanks very much :+1: 

At the moment I'm working on a data set which is a mixture of fitbit (accelerometer, battery, light, sleep depth, steps, ..) data, and symptom score and sleep log questionnaire data. It's from a small pilot study (n=16, 8 weeks), to try to detect early [signs of depressive relapse in schizophrenia patients](https://kclpure.kcl.ac.uk/portal/files/66027798/SleepSight_A_Wearables_KERZ_Published12September2016_GREEN_AAM.pdf). There's a 15 min video presentation here [05:00-23:00](https://skillsmatter.com/skillscasts/10009-data-vis-in-healthcare). 

Because it was a short study we don't really have any "labels" of relapse events, so I'm not really sure how to train a classifier - i.e. what the right loss function is in this case? 

Some of the data from the fitbit, as well as the various smartphones assays is missing, so at the moment we're using Kalman smoothing for imputation. I'm guessing though "padding" would be a better way of dealing with the missing data/asynchronicity? The time-series for the fitbit heart rate data is 1-min granularity, so at it's finest it's typically about 70,000 data points, and there's about 50-60 questionnaires for each participant. Anecdotally, the probability of a relapse event is compounded from successive nights of progressively worsening sleep quality, and deterioration from normative daytime behaviour e.g. lowered activity levels, social interaction, cognitive tests, .. So I guess I need to experiment with the right sampling frequencies? In the absence of "labels" of definite relapse events, prediction of symptom questionnaire scores might be a reasonable problem formulation to try?    

The prepossessing code's [open source](https://github.com/KHP-Informatics/sleepsight-analytics/tree/master/preprocessing), and I'll try to put everything together in a single repository, once I get it all to play nicely together.

I'm not really expecting much from this dataset though. At the moment I'd be really very happy simply to understand your paper and code well, and think of ways how to adapt it for relapse event prediction. I think mathematically it could be formulated as a change point prediction problem, but I'm still working on that? 

There's three much larger/longer studies (n=500, 12 months, x3) which will begin soon - the projects called [RADAR-CNS](https://www.radar-cns.org/strategy). The canonical problem of relapse event prediction is fairly similar for all the three specific medical applications, and the short video above is probably the simplest explanation. I can send you more details if you're interested? I'm guessing from listening to a recent interview of [Katherine Heller](http://www.thetalkingmachines.com/blog/2017/7/30/the-church-of-bayes-and-collecting-data) that your teams working on similar problems, so if I can be of any help just give me a shout :)

Thanks very much once again for sharing your code :+1:

**Comment by jfutoma** [comment]
No problem! Will clean up the code soon so that you can actually run a working example - right now the code I dumped is a bit of a mess, so not sure how useful it is in its current state.

Seems like a hard problem as well.  Labels in our case are also a bit ambiguous (there's no exact "time of sepsis" that's super easy to identify, so we have to come up with a way to create labels).  But, this method as it stands really only directly applies to labelled data, so not sure how you'd directly apply to the unlabeled setting.  Labeling with your questionnaire data might be a reasonable way to do that - it's super easy to adapt the RNN to do multi-class classification where there are multiple labels (in my setting everything was binary: sepsis or not sepsis).

I've thought a bit about change point prediction for other projects I've worked on, and that's definitely a nice way to try to frame the problem when you don't have labels at all.

If it would be at all helpful, shoot me an email and I'm happy to share my slides from ICML, and the video from my talk there should be online soon.  I've also done more recent work extending this ICML model in different directions, paper here (http://mucmd.org/CameraReadySubmissions/53%5CCameraReadySubmission%5CCR.pdf) and soon to be on Arxiv.  You might want to get a better grasp on the ICML paper first before you spend too much time on this one though :) 

Re sample sizes - I was just asking because the way the MGPs are currently implemented it's going to be tricky to scale to absolutely massive data, so you may need to tweak with the granularity of both the input data and the output data on the evenly spaced grid.  How much missing data do you have - is that a big problem for you?  A main motivation for this work is that a lot of clinical time series data is very uncertain and has lots of missing values, and the MGP is pretty good at handling that uncertainty and this framework here maintains uncertainty in the downstream classifier.

Hope this is helpful!
